rewrite-core/src/main/java/org/openrewrite/GitRemote.java

```diff
@@ -129,66 +177,134 @@ public GitRemote parse(String url) {
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElseGet(() -> {
+                        String[] segments = normalizedUri.getPath().split("/");
+                        String origin1 = normalizedUri.getHost() + maybePort(normalizedUri.getPort(), normalizedUri.getScheme());
+                        if (segments.length > 2) {
+                            origin1 += Arrays.stream(segments, 0, segments.length - 2).collect(Collectors.joining("/"));
+                        }
+                        return new RemoteServerMatch(Service.Unknown, origin1, URI.create(normalizedUri.getScheme() + "://" + origin1));
+                    });
+        }
+
+        private String repositoryPath(RemoteServerMatch match, URI normalizedUri) {
+            URI origin = match.matchedUri;
+            String uri = normalizedUri.toString();
+            String contextPath = origin.getPath();
+            String path = normalizedUri.getPath();
+            if (!normalizedUri.getHost().equals(origin.getHost()) || normalizedUri.getPort() != origin.getPort() || !path.startsWith(contextPath)) {
+                throw new IllegalArgumentException("Origin: " + origin + " does not match the clone url: " + uri);
             }
+            return path.substring(contextPath.length())
+                    .replaceFirst("^/", "");
+        }
 
-            private String concat() {
-                StringBuilder builder = new StringBuilder(64);
-                if (host != null) {
-                    builder.append(host);
-                }
-                if (!isDefaultPort()) {
-                    builder.append(':').append(port);
+        private static final Pattern PORT_PATTERN = Pattern.compile(":\\d+");
+
+        static URI normalize(String url) {
+            try {
+                URIish uri = new URIish(url);
+                String scheme = uri.getScheme();
+                String host = uri.getHost();
+                if (host == null) {
+                    if (scheme == null) {
+                        if (url.contains(":")) {
+                            // Looks like SCP style url
+                            scheme = "ssh";
+                        } else {
+                            scheme = "https";
+                        }
+                        uri = new URIish(scheme + "://" + url);
+                        host = uri.getHost();
+                    } else if (!"file".equals(scheme)) {
+                        throw new IllegalStateException("No host in url: " + url);
+                    }
                 }
-                if (!path.isEmpty()) {
-                    if (builder.length() != 0) {
-                        builder.append('/');
+                if (scheme == null) {
+                    if (PORT_PATTERN.matcher(url).find()) {
+                        throw new IllegalArgumentException("Unable to normalize URI. Port without a scheme is not supported: " + url);
```

Comment:
nit: `"Unable to normalize URL: Specifying a port without a scheme is not supported for URL " + url`